src/main/java/org/cactoos/io/AtomicFile.java

```diff
@@ -0,0 +1,70 @@
+/**
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2017-2018 Yegor Bugayenko
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package org.cactoos.io;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.Charset;
+
+/**
+ * A file that can only be written to in an atomic way.
+ *
+ * <p>
+ * There is no thread-safety guarantee.
+ *
+ * @author Ashton Hogan (https://twitter.com/TheAshtonHogan)
+ * @version $Id$
+ * @since 0.49.2
+ */
+public final class AtomicFile extends File {
+
+    private static final long serialVersionUID = -276474606113419118L;
+
+    /**
+     * Ctor.
+     *
+     * @param pathname that should be written to
+     */
+    public AtomicFile(final String pathname) {
+        super(pathname);
+    }
+
+    public synchronized void rewrite(final String string, final Charset charset) throws IOException {
+        File tempFile = new File(System.getProperty("java.io.tmpdir") + File.separator + this.getName() + "_tmp");
+        if ((!this.exists()) && (!tempFile.exists())) {
+            throw new IOException(this.getClass().getName() + " could not perform write. File and Temp File do not exist:\n" + "File: " + this.getAbsolutePath() + "\n" + "Temp File: " + tempFile.getAbsolutePath());
+        }
+        if ((!this.exists()) && (tempFile.exists())) {
+            tempFile.renameTo(this);
+        }
+        if ((this.exists()) && (tempFile.exists())) {
+            tempFile.delete();
+        }
+        OutputTo outputTo = new OutputTo(tempFile, Boolean.TRUE);
+        outputTo.stream().write(string.getBytes(charset));
+        this.delete();
+        tempFile.renameTo(this);
+    }
```

Comment:
@ashtonhogan I don't understand the logic here at all: a temp file is created along ``this`` file, then check if they both exist; if not, exception. 

If ``this`` does not exist, then rename temp to ``this``'s name. 

If both of them exist, then delete the temp file? 

Finally, write the given contents to that outputstream, delete ``this`` and rename ``tempFile`` to ``this``'s name? 

It seems it just juggles with these 2 files, but where is the guarantee atomicity?

Reply:
1. First `if`: Checks to see if there are any files to work with, if not then it throws an exception

2. Second `if`: Finished writing to the temp file, deleted the original then got interrupted

3. Third `if`: Previous write operation did not complete before getting interrupted

Now that any previous interrupts have been dealt with, let's create a temp file, write to it, delete the original and rename the temp file

Reply:
What if it enters the 3rd if? ``tempFile`` will be deleted, then in the very next lines, it will try to write to it and it will pop because tempFile is not there anymore, right?

Reply:
And does renaming change the whole path? Or just the file's name? What if they are in different locations?

Reply:
I made an assumption there, I thought since `OutputTo` does `mkdirs` it might also create the file:

```
public OutputTo(final File file, final boolean mkdirs) {
        this(
            () -> {
                if (mkdirs) {
                    file.getParentFile().mkdirs();
                }
                return new FileOutputStream(file);
            }
        );
    }
```

And maybe it should otherwise `mkdirs` doesn't really make sense?